src/metric.rs

```diff
@@ -0,0 +1,305 @@
+pub enum Semamtics {
+    Counter  = 1,
+    Instant  = 3,
+    Discrete = 4
+}
+
+pub struct Metric<T> {
+    name: String,
+    item: u32,
+    sem: Semamtics,
+    indom: u32,
+    unit: u32,
+    shorthelp: String,
+    longhelp: String,
+    val: T,
+    mmap_view: MmapViewSync
+}
+
+lazy_static! {
+    static ref SCRATCH_VIEW: MmapViewSync = {
+        Mmap::anonymous(super::STRING_BLOCK_LEN as usize, Protection::ReadWrite).unwrap()
+            .into_view_sync()
+    };
+}
+
+impl<T: MetricType + Clone> Metric<T> {
+    pub fn new(
+        name: &str, item: u32, sem: Semamtics,
+        unit: Unit, init_val: T,
+        shorthelp: &str, longhelp: &str) -> Result<Self, String> {
+        
+        if name.len() >= super::METRIC_NAME_MAX_LEN as usize {
+            return Err(format!("name longer than {} bytes", super::METRIC_NAME_MAX_LEN - 1));
+        }
+        if shorthelp.len() >= super::STRING_BLOCK_LEN as usize {
+            return Err(format!("short help text longer than {} bytes", super::STRING_BLOCK_LEN - 1));
+        }
+        if longhelp.len() >= super::STRING_BLOCK_LEN as usize {
+            return Err(format!("long help text longer than {} bytes", super::STRING_BLOCK_LEN - 1));
+        }
+
+        Ok(Metric {
+            name: name.to_owned(),
+            item: item & ((1 << ITEM_BIT_LEN) - 1),
+            sem: sem,
+            indom: 0,
+            unit: unit.pmapi_repr(),
+            shorthelp: shorthelp.to_owned(),
+            longhelp: longhelp.to_owned(),
+            val: init_val,
+            mmap_view: unsafe { SCRATCH_VIEW.clone() }
+        })
+    }
+
+    pub fn val(&self) -> T {
+        self.val.clone()
+    }
+
+    pub fn set_val(&mut self, new_val: T) -> io::Result<()> {
+        new_val.write_to_writer(unsafe { &mut self.mmap_view.as_mut_slice() })?;
+        self.val = new_val;
+        Ok(())
+    }
+}
+
+pub trait MMVMetric {
+    fn name(&self) -> &str;
+    fn item(&self) -> u32;
+    fn type_code(&self) -> u32;
+    fn sem(&self) -> &Semamtics;
+    fn unit(&self) -> u32;
+    fn indom(&self) -> u32;
+    fn shorthelp(&self) -> &str;
+    fn longhelp(&self) -> &str;
+    fn write_value(&mut self, cursor: &mut Cursor<&mut [u8]>) -> io::Result<()>;
+    fn set_mmap_view(&mut self, mmap_view: MmapViewSync);
+}
+
+impl<T: MetricType> MMVMetric for Metric<T> {
+    fn name(&self) -> &str { &self.name }
+
+    fn item(&self) -> u32 { self.item }
+
+    fn type_code(&self) -> u32 { self.val.type_code() }
+
+    fn sem(&self) -> &Semamtics { &self.sem }
+
+    fn unit(&self) -> u32 { self.unit }
+
+    fn indom(&self) -> u32 { self.indom }
+
+    fn shorthelp(&self) -> &str { &self.shorthelp }
+
+    fn longhelp(&self) -> &str { &self.longhelp }
+
+    fn write_value(&mut self, cursor: &mut Cursor<&mut [u8]>) -> io::Result<()> {
+        self.val.write_to_writer(cursor)
+    }
+
+    fn set_mmap_view(&mut self, mmap_view: MmapViewSync) {
+        self.mmap_view = mmap_view;
+    }
+}
```

Comment:
My idea is that we push all MMV data structures and methods to write/dump MMV files into an `mmv` module so we can separate all that from the high-level instrumentation specific parts.